snuba/query/validation/validators.py

```diff
@@ -0,0 +1,88 @@
+from abc import ABC, abstractmethod
+from typing import Any, Optional, Sequence, Set
+
+from snuba.query import Query
+from snuba.query.conditions import (
+    ConditionFunctions,
+    get_first_level_and_conditions,
+)
+from snuba.query.exceptions import InvalidQueryException
+from snuba.query.expressions import Expression
+from snuba.query.matchers import Any as AnyMatch
+from snuba.query.matchers import AnyOptionalString
+from snuba.query.matchers import Column as ColumnMatch
+from snuba.query.matchers import FunctionCall as FunctionCallMatch
+from snuba.query.matchers import Literal as LiteralMatch
+from snuba.query.matchers import Or
+from snuba.query.matchers import String as StringMatch
+
+
+class QueryValidator(ABC):
+    """
+    Contains validation logic that requires the entire query. An entity has one or more
+    of these validators that it adds contextual information too.
+    """
+
+    @abstractmethod
+    def validate(self, query: Query, alias: Optional[str] = None,) -> None:
+        """
+        Validate that the query is correct. If the query is not valid, raise an
+        Exception, otherwise return None. If the entity that calls this is part
+        of a join query, the alias will be populated with the entity's alias.
+
+        :param query: The query to validate.
+        :type query: Query
+        :param alias: The alias of the entity in a JOIN query.
+        :type alias: Optional[str]
+        :raises InvalidQueryException: [description]
+        """
+        raise NotImplementedError
+
+
+def build_match(
```

Comment:
This seems a fairly generic and reusable method, why don't we move it to `query/conditions.py`